test/generators/javascript/preset/__snapshots__/MarshallingPreset.spec.ts.snap

```diff
@@ -0,0 +1,165 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`Marshalling preset should render un/marshal code 1`] = `
```

Comment:
I like to take the snapshots and ensure they perform the needed functionality, i.e. copy-paste it into a JS environment in this case. As we have a bit of a problem automating this to so many languages 😟 

However, snapshots can be a bit hard to figure out the actual output, as it escapes characters which you need to replace: 
- \` to `
- \$ to $
- \\" to "

<details>
<summary>Here is the cleaned up version</summary>
<p>

```js
class Test {
  stringProp;
  numberProp;
  objectProp;
  additionalProperties;
  sTestPatternProperties;
  sAnotherTestPatternProperties;
  constructor(input) {
    this.stringProp = input.stringProp;
    if (input.hasOwnProperty('numberProp')) {
      this.numberProp = input.numberProp;
    }
    if (input.hasOwnProperty('objectProp')) {
      this.objectProp = input.objectProp;
    }
  }
  get stringProp() { return this.stringProp; }
  set stringProp(stringProp) { this.stringProp = stringProp; }
  get numberProp() { return this.numberProp; }
  set numberProp(numberProp) { this.numberProp = numberProp; }
  get objectProp() { return this.objectProp; }
  set objectProp(objectProp) { this.objectProp = objectProp; }
  get additionalProperties() { return this.additionalProperties; }
  set additionalProperties(additionalProperties) { this.additionalProperties = additionalProperties; }
  get sTestPatternProperties() { return this.sTestPatternProperties; }
  set sTestPatternProperties(sTestPatternProperties) { this.sTestPatternProperties = sTestPatternProperties; }
  get sAnotherTestPatternProperties() { return this.sAnotherTestPatternProperties; }
  set sAnotherTestPatternProperties(sAnotherTestPatternProperties) { this.sAnotherTestPatternProperties = sAnotherTestPatternProperties; }
  public marshal(): string {
    let json = '{'
    if (this.stringProp !== undefined) {
      json += `"string prop": ${typeof this.stringProp === 'number' || typeof this.stringProp === 'boolean' ? this.stringProp : JSON.stringify(this.stringProp)},`;
    }
    if (this.numberProp !== undefined) {
      json += `"numberProp": ${typeof this.numberProp === 'number' || typeof this.numberProp === 'boolean' ? this.numberProp : JSON.stringify(this.numberProp)},`;
    }
    if (this.objectProp !== undefined) {
      json += `"objectProp": ${this.objectProp.marshal()},`;
    }
    if (this.sTestPatternProperties !== undefined) {
      for (const [key, value] of this.sTestPatternProperties.entries()) {
        //Only render pattern properties which are not already a property
        if (Object.keys(this).includes(String(key))) continue;
        json += `"${key}": ${typeof value === 'number' || typeof value === 'boolean' ? value : JSON.stringify(value)},`;
      }
    } if (this.sAnotherTestPatternProperties !== undefined) {
      for (const [key, value] of this.sAnotherTestPatternProperties.entries()) {
        //Only render pattern properties which are not already a property
        if (Object.keys(this).includes(String(key))) continue;
        json += `"${key}": ${value.marshal()},`;
      }
    }
    if (this.additionalProperties !== undefined) {
      for (const [key, value] of this.additionalProperties.entries()) {
        //Only render additionalProperties which are not already a property
        if (Object.keys(this).includes(String(key))) continue;
        json += `"${key}": ${value.marshal()},`;
      }
    }
    //Remove potential last comma 
    return `${json.charAt(json.length - 1) === ',' ? json.slice(0, json.length - 1) : json}}`;
  }
  public static unmarshal(json: string | object): Test {
    const obj = typeof json === "object" ? json : JSON.parse(json);
    const instance = new Test({});
    if (obj["string prop"] !== undefined) {
      instance.stringProp = obj["string prop"];
    }
    if (obj["numberProp"] !== undefined) {
      instance.numberProp = obj["numberProp"];
    }
    if (obj["objectProp"] !== undefined) {
      instance.objectProp = NestedTest.unmarshal(obj["objectProp"]);
    }
    //Not part of core properties
    if (instance.sTestPatternProperties === undefined) { instance.sTestPatternProperties = new Map(); }
    if (instance.sAnotherTestPatternProperties === undefined) { instance.sAnotherTestPatternProperties = new Map(); }
    if (instance.additionalProperties === undefined) { instance.additionalProperties = new Map(); }
    for (const [key, value] of Object.entries(obj).filter((([key,]) => { return !["string prop", "numberProp", "objectProp"].includes(key); }))) {
      //Check all pattern properties
      if (key.match(new RegExp('^S(.?)test'))) {
        instance.sTestPatternProperties.set(key, value as any);
        continue;
      }//Check all pattern properties
      if (key.match(new RegExp('^S(.?)AnotherTest'))) {
        instance.sAnotherTestPatternProperties.set(key, NestedTest.unmarshal(value as any));
        continue;
      }
      instance.additionalProperties.set(key, NestedTest.unmarshal(value));
    }
    return instance;
  }
}
class NestedTest {
  stringProp;
  additionalProperties;
  constructor(input) {
    if (input.hasOwnProperty('stringProp')) {
      this.stringProp = input.stringProp;
    }
  }
  get stringProp() { return this.stringProp; }
  set stringProp(stringProp) { this.stringProp = stringProp; }
  get additionalProperties() { return this.additionalProperties; }
  set additionalProperties(additionalProperties) { this.additionalProperties = additionalProperties; }
  public marshal(): string {
    let json = '{'
    if (this.stringProp !== undefined) {
      json += `"stringProp": ${typeof this.stringProp === 'number' || typeof this.stringProp === 'boolean' ? this.stringProp : JSON.stringify(this.stringProp)},`;
    }

    if (this.additionalProperties !== undefined) {
      for (const [key, value] of this.additionalProperties.entries()) {
        //Only render additionalProperties which are not already a property
        if (Object.keys(this).includes(String(key))) continue;
        json += `"${key}": ${typeof value === 'number' || typeof value === 'boolean' ? value : JSON.stringify(value)},`;
      }
    }
    //Remove potential last comma 
    return `${json.charAt(json.length - 1) === ',' ? json.slice(0, json.length - 1) : json}}`;
  }
  public static unmarshal(json: string | object): NestedTest {
    const obj = typeof json === "object" ? json : JSON.parse(json);
    const instance = new NestedTest({});
    if (obj["stringProp"] !== undefined) {
      instance.stringProp = obj["stringProp"];
    }
    //Not part of core properties

    if (instance.additionalProperties === undefined) { instance.additionalProperties = new Map(); }
    for (const [key, value] of Object.entries(obj).filter((([key,]) => { return !["stringProp"].includes(key); }))) {

      instance.additionalProperties.set(key, value);
    }
    return instance;
  }
}
```

</p>
</details>  

Throw that code into a new JS environment and try testing it manually, also to see if it has some problems with syntax 🙂 That way you can go back to the preset and adapt it so provide you with the correct output. 
 
```js
const instance = new Test({stringProp: 'test'});
const json = instance.marshal();
const unmarshalledInstance = Test.unmarshal(json);
```